### PR TITLE
Integrate pytest-cov into `azdev test`

### PR DESCRIPTION
**Pull request by StrawnSC**
This change would allow CLI devs to generate coverage.py-style reports on test coverage with `azdev test --coverage`. Hopefully it will help CLI devs write higher quality tests for the CLI command modules. 

Running `azdev test --coverage ...` runs the specified tests (as usual), then prints a brief test coverage report of [this form](https://pytest-cov.readthedocs.io/en/latest/readme.html#usage). By default (without the `--no-htmlcov` argument), the command also generates a static webpage in the `htmlcov/` directory. The site is an interactive coverage report, showing file-by-file and line-by-line test coverage. [Here](https://nedbatchelder.com/files/sample_coverage_html/index.html) is a sample html report. 

Besides the arguments already listed, there are three others related to coverage: 
`--append-coverage` adds coverage data to any preexisting coverage data, allowing one to aggregate coverage across multiple CLI calls. 
`--coverage-path <path>` restricts the coverage to only track source files in the given directory. This can be useful for limiting coverage to just one CLI command module. By default, the coverage path is the azure CLI path set with `azdev setup`
`--open-coverage` opens the html coverage report in the default browser when the command finishes 

See [here](https://coverage.readthedocs.io/en/coverage-5.5/) and [here](https://pytest-cov.readthedocs.io/en/latest/readme.html#usage) for more info on coverage.py and pytest-cov, respectively

### CODE REVIEW

**Comment by wangzelin007** [comment]
Hi @StrawnSC @jiasli @kairu-ms @jsntcy 
I was wondering can we put the `--coverage` out of `azdev test` command to become a independent command like `azdev coverage`.
Because running `azdev test` is very time wasting, but only `azdev coverage` is very fast.
If I only want to generate the coverage report, I need to spend a lot of time waiting for the `azdev test`.
And a failed test will cause the coverage test terminate unexpected.
This is my experience of downloading the code and actually using it.

File: .gitignore [hunk 1]
```diff
@@ -12,3 +13,6 @@ azdev.egg-info/
 pip-wheel-metadata/*
 build/
 dist/
+.DS_Store
+htmlcov/
+.coverage
```
Comment (wangzelin007) [inline .gitignore]: Please add a newline at end of file, others LGTM

**Comment by StrawnSC** [comment]
@wangzelin007 thanks for taking a look at the PR, I really appreciate you trying it out! To address your comments: 
> I was wondering can we put the `--coverage` out of `azdev test` command to become a independent command like `azdev coverage`.
> Because running `azdev test` is very time wasting, but only `azdev coverage` is very fast.
> If I only want to generate the coverage report, I need to spend a lot of time waiting for the `azdev test`.

I'm not sure we would be able to do it like this. The way that the code coverage tool works, it needs to be run while tests are being executed. This is how it can generate reports on which lines of code were executed during testing. (See the tool's documentation [here](https://coverage.readthedocs.io/en/6.1.1/) and [here](https://pytest-cov.readthedocs.io/en/latest/) for more info). I don't believe it can be used as a static analysis tool
Unfortunately this means that getting a coverage report will be pretty slow (since testing can be slow and the coverage tool introduces some overhead), but I imagine that the user wouldn't need to run it every time they test. I think the typical use case would be the user running `azdev test --coverage ...` after writing tests but before opening a PR, just to be sure that their tests fully cover the new functionality. They would probably run it only on a subset of tests as well.

> And a failed test will cause the coverage test terminate unexpected.

I don't think I've tested it against a failed test so I'll take a look at that behavior and see if I can fix it

**Comment by wangzelin007** [comment]
@StrawnSC 
Yes, you are right.
Usually coverage is fast because it runs with unit tests, but our tests almost are e2e tests.
With reference to the second question, the interruption is caused by the original logic of the azdev test, and can use `azdev test -c --no-exitfirst` to forbid exiting on error or failed test.

**Review by wangzelin007 — APPROVED** [review]

**Comment by StrawnSC** [comment]
@wangzelin007 thanks again for taking a look at this. Can I merge this PR? It looks like there aren't regular milestones / release dates for this repo

**Comment by StrawnSC** [comment]
Hey @wangzelin007 just wanted to follow up on this again. It'd be useful for my CLI work to get this merged in soon. Can I merge this PR or does someone on the core CLI team need to do that?

**Comment by wangzelin007** [comment]
@StrawnSC Sorry for the late reply, actually our team has a discussion about this PR，and we think CLI developer care more about the command level coverage other than the code level coverage.
I already develop a new feature allow CLI dev tools to atuo generate CLI Command Coverage Report about every module to track and improve command-level test coverage. (done)
You can check the [PR](https://github.com/Azure/azure-cli-dev-tools/pull/323) here.
We are going to release this new feature next month.
And we also plan to add a new rule in linter check to detect any missing test cases in every pull request, so we can continuously improve code quality. ( in progress)
Please confirm whether this function can meet your testing requirements?

**Comment by StrawnSC** [comment]
Hey @wangzelin007, thanks for the explanation -- sorry for getting back to you so late. I didn't take the chance to check this out before going OOF

I'm not sure that the new feature fulfills the same role that this feature would. There were a few bugs in the `appservice` command module that were caused by not having tests cover certain portions of code within a command. For example, we would have tests using a command, but that missed part of the command under an 'if' statement. Would your feature be able to pick up on cases like that? If not, would the core CLI team be opposed to having support for both statement coverage (this PR) and command coverage (your PR) in azdev?

I was going to try out your the command coverage feature, but ran into an issue ([see here](https://github.com/Azure/azure-cli-dev-tools/pull/323#issuecomment-1004311297))

**Comment by wangzelin007** [comment]
@StrawnSC,
It makes sense. 
My PR can not support the `if` statement cases.
I will discuss with my team today, hope we can release a new version for this two PR.

**Comment by jiasli** [comment]
The right process to onboard a new feature is to contact azpycli@microsoft.com and Azure CLI team's manager and product manager for reviewing and planning.

Unapproved feature can lead to instability in production code and maintenance cost.

This PR should be reverted.

**Pull request merged**